Flaky TestRuler

The `TestRuler` is flaky ([failed CI run](https://github.com/grafana/mimir/runs/5022450868?check_suite_focus=true)):

```
level=info msg="ring doesn't exist in KV store yet"
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler up and running"
level=error msg="failed to check if the ruler replica owns the rule group" user=user1 namespace=)(_+?/|namespace1+/? group=)(_+?/|group1+/? err="error reading ring to verify rule group ownership: empty ring"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
--- FAIL: TestRuler (0.06s)
    --- FAIL: TestRuler/rules_special_characters (0.01s)
        api_test.go:190: 
            	Error Trace:	api_test.go:190
            	Error:      	Not equal: 
            	            	expected: "{\"status\":\"success\",\"data\":{\"groups\":[{\"name\":\")(_+?/|group1+/?\",\"file\":\")(_+?/|namespace1+/?\",\"rules\":[{\"name\":\"UP_RULE\",\"query\":\"up\",\"labels\":{},\"health\":\"unknown\",\"lastError\":\"\",\"type\":\"recording\",\"lastEvaluation\":\"0001-01-01T00:00:00Z\",\"evaluationTime\":0},{\"state\":\"inactive\",\"name\":\"UP_ALERT\",\"query\":\"up \\u003c 1\",\"duration\":0,\"labels\":{},\"annotations\":{},\"alerts\":[],\"health\":\"unknown\",\"lastError\":\"\",\"type\":\"alerting\",\"lastEvaluation\":\"0001-01-01T00:00:00Z\",\"evaluationTime\":0}],\"interval\":60,\"lastEvaluation\":\"0001-01-01T00:00:00Z\",\"evaluationTime\":0,\"sourceTenants\":null}]},\"errorType\":\"\",\"error\":\"\"}"
            	            	actual  : "{\"status\":\"success\",\"data\":{\"groups\":[]},\"errorType\":\"\",\"error\":\"\"}"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-{"status":"success","data":{"groups":[{"name":")(_+?/|group1+/?","file":")(_+?/|namespace1+/?","rules":[{"name":"UP_RULE","query":"up","labels":{},"health":"unknown","lastError":"","type":"recording","lastEvaluation":"0001-01-01T00:00:00Z","evaluationTime":0},{"state":"inactive","name":"UP_ALERT","query":"up \u003c 1","duration":0,"labels":{},"annotations":{},"alerts":[],"health":"unknown","lastError":"","type":"alerting","lastEvaluation":"0001-01-01T00:00:00Z","evaluationTime":0}],"interval":60,"lastEvaluation":"0001-01-01T00:00:00Z","evaluationTime":0,"sourceTenants":null}]},"errorType":"","error":""}
            	            	+{"status":"success","data":{"groups":[]},"errorType":"","error":""}
            	Test:       	TestRuler/rules_special_characters
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ruler using enabled users" enabled=user1
level=info msg="ruler using disabled users" disabled=user1
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler using disabled users" disabled=user1
level=info msg="ruler using disabled users" disabled=user1
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ring doesn't exist in KV store yet"
level=info msg="instance not found in the ring" instance=localhost ring=ruler
level=info msg="ring doesn't exist in KV store yet"
level=info msg="ruler up and running"
level=info msg="ring lifecycler is shutting down" ring=ruler
level=info msg="unregistering instance from ring" ring=ruler
level=info msg="instance removed from the ring" ring=ruler
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_disabled_with_a_federated_rules_and_a_regular_rule800878704/001
level=info msg="updating rule file" file=/tmp/TestRuler_TenantFederationFlagtenant_federation_disabled_with_a_federated_rules_and_a_regular_rule800878704/001/tenant-1/ns
level=info user=tenant-1 msg="Stopping notification manager..."
level=info msg="stopping user managers"
level=info user=tenant-1 msg="Stopping rule manager..."
level=info user=tenant-1 msg="Rule manager stopped"
level=info msg="all user managers stopped"
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_disabled_with_a_federated_rules_and_a_regular_rule800878704/001
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_only_with_federated_groups1584402535/001
level=info msg="updating rule file" file=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_only_with_federated_groups1584402535/001/tenant-1/ns
level=info user=tenant-1 msg="Stopping notification manager..."
level=info msg="stopping user managers"
level=info user=tenant-1 msg="Stopping rule manager..."
level=info user=tenant-1 msg="Rule manager stopped"
level=info msg="all user managers stopped"
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_only_with_federated_groups1584402535/001
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_with_federated_and_regular_groups3480484530/001
level=info msg="updating rule file" file=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_with_federated_and_regular_groups3480484530/001/tenant-1/ns
level=info user=tenant-1 msg="Stopping notification manager..."
level=info msg="stopping user managers"
level=info user=tenant-1 msg="Stopping rule manager..."
level=info user=tenant-1 msg="Rule manager stopped"
level=info msg="all user managers stopped"
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_enabled_with_federated_and_regular_groups3480484530/001
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_disabled730375253/001
level=info user=tenant-1 msg="Stopping notification manager..."
level=info msg="stopping user managers"
level=info user=tenant-1 msg="Stopping rule manager..."
level=info user=tenant-1 msg="Rule manager stopped"
level=info msg="all user managers stopped"
level=info msg="cleaning up mapped rules directory" path=/tmp/TestRuler_TenantFederationFlagtenant_federation_disabled730375253/001
FAIL
FAIL	github.com/grafana/mimir/pkg/ruler	9.023s
```

--- a/pkg/ruler/ruler.go
+++ b/pkg/ruler/ruler.go
@@ -325,6 +325,15 @@ func (r *Ruler) starting(ctx context.Context) error {
 		return errors.Wrap(err, "unable to start ruler subservices")
 	}
 
+	// Wait until the ring client detected this instance in the ACTIVE state to
+	// make sure that when we'll run the initial sync we already know  the tokens
+	// assigned to this instance.
+	level.Info(r.logger).Log("msg", "waiting until ruler is ACTIVE in the ring")
+	if err := ring.WaitInstanceState(ctx, r.ring, r.lifecycler.GetInstanceID(), ring.ACTIVE); err != nil {
+		return err
+	}
+	level.Info(r.logger).Log("msg", "ruler is ACTIVE in the ring")
+
 	// TODO: ideally, ruler would wait until its queryable is finished starting.
 	return nil
 }

